server/tserver/src/main/java/org/apache/accumulo/tserver/log/LogSorter.java

```diff
@@ -274,12 +274,7 @@ void writeBuffer(String destPath, List<Pair<LogFileKey,LogFileValue>> buffer, in
       var logFileKey = pair.getFirst();
       var logFileValue = pair.getSecond();
       Key k = logFileKey.toKey();
-      var list = keyListMap.putIfAbsent(k, logFileValue.mutations);
-      if (list != null) {
-        var muts = new ArrayList<>(list);
-        muts.addAll(logFileValue.mutations);
-        keyListMap.put(logFileKey.toKey(), muts);
-      }
+      keyListMap.computeIfAbsent(k, (key) -> new ArrayList<>()).addAll(logFileValue.mutations);
```

Comment:
Seems like this code was attempting to avoid any copy when the key was only seen once.  However when a key was seen more than once it would keep copying.   Wondering if we can efficiently accomplish the following in some way.

1. The first time a key is seen put a reference to the list in the map w/o copy
2. The second time a key is seen, copy the list and then append
3. The third or later time a keys is seen append to list and do not copy.

I was able to create a test that reproduces this problem.  Will post that in another cmment and try this change w/ the test.

Reply:
Good point about not copying in the first case. The code was optimal when the key was only seen once.

Reply:
This change definitely fixes the problem.  Sorting a log w/ 20M key/values stored in 20K log entries w/ the same log file key took 10s.  Before this change it took so long that  I killed the test instead of waiting for it.

Reply:
Do we need to optimize for the single key case so as not to copy the mutations?

Reply:
> Do we need to optimize for the single key case so as not to copy the mutations?

Not sure, looking into this.

Reply:
No we do not need that optimization.  The key in this case is not related to the data, but is related to the tablet and minor compaction count. So for each tablet in the walog it will have the same key until a minor compactoin happens at which point the key changes.  Since the key is per tablet in walog,  does not seem that further optimization to avoid the initial copy is worthwhile.

Looking into this I realized the problem is more general than what was described in the issue.  Can run into this problem even if writing different keys to the same tablet.  The problem is more if a function of how many entries an individual tablet has in the walog, does not really matter what the data in the tablet is.